For EK60 and EK80 add `beam` and `ping_time` to `Beam_groupX`

b-reyes opened the pull request:
This PR addresses the addition of the dimensions `beam` and `ping_time` to variables within `Beam_groupX`. The specific variables were identified in #520. Note that this PR is only concerned with the sensors EK60 and EK80. The modifications necessary for AZFP will take place in a different PR. 

To complete this task, I have added the function `beamgroups_to_convention()` in `echopype/convert/set_groups_base.py`. This function depends on several dictionaries I have defined at the top of the file. This function is called at the bottom of `set_beam()` in the `set_groups_ekXX.py` files. Note: currently, my intention is to use this function again when I create my code for the conversion from v0.5.x to v0.6.x in `open_converted`.

leewujung commented:
@b-reyes : this is awesome!! The code is very clean!! 😍   My head is not clear enough to review this right now so it has to be tomorrow, but I just want to say thank you for putting so much time and thought into this (and the gigantic #520 discussion thread 🤯).

leewujung commented on echopype/convert/set_groups_base.py:
```diff
@@ -11,6 +11,58 @@
 DEFAULT_CHUNK_SIZE = {"range_sample": 25000, "ping_time": 2500}
 
+# Variables that need only the beam dimension added to them.
+# These lists include all Sonar beam groups.
```
```suggestion
# These lists are applied to all Sonar/Beam_groupX groups.
```
Just very minor wording suggestion as clarification.

leewujung commented on echopype/convert/set_groups_base.py:
```diff
@@ -210,3 +262,101 @@ def _beam_groups_vars(self):
+                        .copy()
+                    )
+            else:
+                # TODO: right now there is no attr or encoding for the beam dimension
+                #  if this changes in the future, we need to add them here.
```
Seems like we can add this in `convention/1.0.yml`?
<img width="678" alt="Screen Shot 2022-04-21 at 11 27 57 AM" src="https://user-images.githubusercontent.com/15334215/164528860-dce89123-550a-4f2c-956d-3574482cbb18.png">

emiliom replied:
Agreed. Add [here](https://github.com/OSOceanAcoustics/echopype/blob/dev/echopype/echodata/convention/1.0.yml#L65). Then you would simply assign attrs using `self._varattrs["beam_coord_default"]["beam"]`

While you're editing 1.0.yml, could you correct the indentation to `valid_min`, [here](https://github.com/OSOceanAcoustics/echopype/blob/dev/echopype/echodata/convention/1.0.yml#L59)? It's supposed to be an attribute of `frequency`, but the indentation is off.

b-reyes replied:
> Agreed. Add [here](https://github.com/OSOceanAcoustics/echopype/blob/dev/echopype/echodata/convention/1.0.yml#L65). Then you would simply assign attrs using `self._varattrs["beam_coord_default"]["beam"]`

Is the following the correct addition (we only need to add the long_name)?

```
    beam:
      long_name: Beam name
```

emiliom replied:
Yeah, that's correct

emiliom commented on echopype/convert/set_groups_base.py:
```diff
@@ -210,3 +262,101 @@ def _beam_groups_vars(self):
+        This does several things:
+        1. Adds ``beam`` dimension to several variables
+        2. Adds ``ping_time`` dimension to several variables
+        """
```
Some suggested edits to clarify what's done and the context (note: I don't know if some of the new line lengths exceed a docstring requirement):

```
Manipulates variables in ``Sonar/Beam_group1``
and ``Sonar/Beam_group2`` to adhere to 
SONAR-netCDF4 vers. 1 with respect to the use of
``ping_time`` and ``beam`` dimensions.

This does several things:
1. Creates ``beam`` dimension and coordinate variable when not present
2. Adds ``beam`` dimension to several variables when missing
3. Adds ``ping_time`` dimension to several variables when missing
```

leewujung reviewed (commented):
Everything looks good! The only other thing I would suggest is to add tests for the dimensions after set_beam, perhaps best done as a separate test module (script) just for the dimensions?

emiliom commented on echopype/convert/set_groups_base.py:
```diff
@@ -210,3 +262,101 @@ def _beam_groups_vars(self):
+            else:
+                # TODO: right now there is no attr or encoding for the beam dimension
+                #  if this changes in the future, we need to add them here.
+                ds[var_name] = (
```
It may be worth adding a clarifying comment here, like this: "Add a single-value beam dimension"

emiliom commented on echopype/convert/set_groups_base.py:
```diff
@@ -210,3 +262,101 @@ def _beam_groups_vars(self):
+
+        # variables to add ping_time to
+        add_ping_time_names = set(ds.variables).intersection(beam_ping_time_names[sonar_model])
+        add_ping_time_names = add_ping_time_names.union(ping_time_only_names[sonar_model])
```
How about using the same single-statement form you used in `_add_beam_dim`?

b-reyes replied:
@emiliom can you clarify this? Are you suggesting that I do the following? 
```
add_ping_time_names = set(ds.variables).intersection(beam_ping_time_names[sonar_model]).union(ping_time_only_names[sonar_model])
```

emiliom replied:
yes

b-reyes replied:
Sounds good. I will go ahead and do that.

emiliom commented on echopype/convert/set_groups_base.py:
```diff
@@ -210,3 +262,101 @@ def _beam_groups_vars(self):
+
+            ds[var_name] = (
+                ds[var_name]
+                .expand_dims(dim={"ping_time": ping_time}, axis=ds[var_name].ndim)
```
I suggest following the cleaner approach you used in `_add_beam_dim`:
```python
.expand_dims(dim={"ping_time": ds.ping_time}, axis=ds[var_name].ndim)
```

emiliom commented:
Are the changes to `calibrate/calibrate_ek.py` unrelated to the thrust of this PR? I don't see an obvious relationship to the addition of beam and ping_time dimensions. It's ok with me if they are in fact unrelated, but please make a note of that in your initial comment

emiliom commented on echopype/convert/set_groups_ek60.py:
```diff
@@ -611,6 +611,9 @@ def set_beam(self) -> xr.Dataset:
             [ds, xr.merge(ds_backscatter)], combine_attrs="override"
         )  # override keeps the Dataset attributes
 
+        # manipulate Dataset to adhere to convention
```
I suggest a more fleshed out comment. Otherwise it sounds generic. Say: 

"Manipulate some Dataset dimensions to adhere to convention"

emiliom commented on echopype/convert/set_groups_ek80.py:
```diff
@@ -659,6 +660,11 @@ def merge_save(ds_combine, ds_type, group_name):
         else:
             ds_beam = merge_save(ds_power, "power", group_name="/Sonar/Beam_group1")
 
+        # manipulate Datasets to adhere to convention
```
I suggest a more fleshed out comment. Otherwise it sounds generic. Say:

"Manipulate some Dataset dimensions to adhere to convention"

emiliom commented:
Actually, I have two suggestions to create some separation of sensor-specific from generic stuff. I think this would work well for the `open_raw` aspects in this PR, but I don't have a good sense of the impacts on your planned `open_converted` changes:

1. Move the `beam_only_names`, `ping_time_only_names` and `beam_ping_time_names` dicts to the `set_groups_SENSOR.py` modules, so they're passed as arguments to `beamgroups_to_convention`.
2. Eliminate the `sonar_model` block in `beamgroups_to_convention`, and instead have the `set_groups_SENSOR.py` modules pass the appropriate sensor to the `beamgroups_to_convention` call.

emiliom commented:
I forgot to comment on downstream impacts on processed data like Sv. It looks like `compute_Sv` (for example) would return an xarray dataset that retains the `beam` dimension. I'm inferring that from the changes @b-reyes made to `test_echodata.py` and `test_calibrate.py`. @leewujung should `compute_Sv` and `compute_TS` be modified so `beam` is squeezed? For EK80 data having both beam and beam power groups, do these functions currently return a dataset with a length-2 `beam` dimension?

b-reyes commented:
> Are the changes to calibrate/calibrate_ek.py unrelated to the thrust of this PR? I don't see an obvious relationship to the addition of beam and ping_time dimensions. It's ok with me if they are in fact unrelated, but please make a note of that in your initial comment

@emiliom these changes are required because of the addition of the dimensions. I have added an edit to my initial comment.

leewujung commented:
Late to the party! 

On the potential blurring between `set_groups_SENSOR.py` and `set_groups_base.py`: one way to resolve this may be to just move those _definitions_ (the `beam_only_names`, `ping_time_only_names` and `beam_ping_time_names` dicts) into another script, e.g. `set_groups_dims_mod.py`. These dicts can then be used by `beamgroups_to_convention`. 

This is along the same line as in our `core.py`, BUT in `core.py` the levels of organization is first with SENSOR and then with specific info, whereas this order is the reversed.

leewujung commented:
> I forgot to comment on downstream impacts on processed data like Sv. It looks like `compute_Sv` (for example) would return an xarray dataset that retains the `beam` dimension. I'm inferring that from the changes @b-reyes made to `test_echodata.py` and `test_calibrate.py`. @leewujung should `compute_Sv` and `compute_TS` be modified so `beam` is squeezed? 

Good question. The main reason is because for EK80 complex data (`Sonar/Beam_group1`) `compute_Sv` would involve a `.mean(dim="beam")` operation that changes the original length=4 `beam` dimension to be of length=1:
https://github.com/OSOceanAcoustics/echopype/blob/bfc22eb1a95c881f766a3253e43a4ed99d90acd8/echopype/calibrate/calibrate_ek.py#L706 and
https://github.com/OSOceanAcoustics/echopype/blob/bfc22eb1a95c881f766a3253e43a4ed99d90acd8/echopype/calibrate/calibrate_ek.py#L726


So, there is a conceptual incongruence and potential confusion that takes place here. This is related to how we use the `beam` dimension to store data from each element of a split-beam transducer.

I wonder if it may be better if we remove the `beam` dimension for the time being given our current scope (scientific echosounder). 

---------------------------

> For EK80 data having both beam and beam power groups, do these functions currently return a dataset with a length-2 `beam` dimension?

No, the complex and power/angle data are calibrated separately in the current `compute_Sv` setup. Users have to pass in both the `waveform_mode` and `encode_mode` to use this function.
https://github.com/OSOceanAcoustics/echopype/blob/bfc22eb1a95c881f766a3253e43a4ed99d90acd8/echopype/calibrate/api.py#L126-L143
Outputs would always just be length=1 along the `beam` dimension.

emiliom commented:
Thanks, @leewujung ! I just realized that this question of mine was non-sensical as written:

> For EK80 data having both beam and beam power groups, do these functions currently return a dataset with a length-2 beam dimension?

I was conflating the existence of 2 **beam groups** with the 4 `beam` coordinate values. Sigh. But I guess your answer both to this bad question and my earlier comments is the same: at this time, `compute_Sv/TS` always return Sv/TS dataarrays with length-1 `beam` dimension. Based on what you've said, **I suggest that we squeeze `beam` out for the returned datasets**. It looks like it can be done easily at the [very end of `_compute_cal`](https://github.com/OSOceanAcoustics/echopype/blob/dev/echopype/calibrate/api.py#L12). It could be applied on the `return` statement itself.

@leewujung can you confirm? If you do, then, @b-reyes please push a new commit with this change. Note that you'll have to undo some of your changes to `test_echodata.py` and `test_calibrate.py`.

b-reyes commented:
> Actually, I have two suggestions to create some separation of sensor-specific from generic stuff. I think this would work well for the `open_raw` aspects in this PR, but I don't have a good sense of the impacts on your planned `open_converted` changes:
> 
> 1. Move the `beam_only_names`, `ping_time_only_names` and `beam_ping_time_names` dicts to the `set_groups_SENSOR.py` modules, so they're passed as arguments to `beamgroups_to_convention`.
> 2. Eliminate the `sonar_model` block in `beamgroups_to_convention`, and instead have the `set_groups_SENSOR.py` modules pass the appropriate sensor to the `beamgroups_to_convention` call.

@emiliom I think this could work. It wouldn't have too much of an effect on my plans for `open_converted`. I would essentially need to have the following in my conversion script: 

```
from ...convert.set_groups_ek60 import SetGroupsEK60
from ...convert.set_groups_ek80 import SetGroupsEK80
from ...convert.set_groups_azfp import SetGroupsAZFP
```
Using `SetGroupsXX` I could then get access to the appropriate dictionary. To make this work though, I would need to make these dictionaries be global variables. One more note, if we do item 1., then there is no need to pass `sonar_model` into `beamgroups_to_convention` (so item 2. would go away).

leewujung commented:
For the beam dimension on outputs of compute_Sv:

> Based on what you've said, **I suggest that we squeeze `beam` out for the returned datasets**. It looks like it can be done easily at the [very end of `_compute_cal`](https://github.com/OSOceanAcoustics/echopype/blob/dev/echopype/calibrate/api.py#L12). It could be applied on the `return` statement itself.
> 
> @leewujung can you confirm? If you do, then, @b-reyes please push a new commit with this change. Note that you'll have to undo some of your changes to `test_echodata.py` and `test_calibrate.py`.

Yes, it is always length=1 and let's just squeeze it out.

emiliom commented:
> Yes, it is always length=1 and let's just squeeze it out.

Thanks! @b-reyes : please go ahead and apply the [changes I've suggested above](https://github.com/OSOceanAcoustics/echopype/pull/638#issuecomment-1105834467). As you'll see, even though the squeeze operation will be implemented in the `_compute_cal` function, it will apply to both `compute_Sv` and `compute_TS`, as expected.

emiliom commented:
> Using SetGroupsXX I could then get access to the appropriate dictionary. To make this work though, I would need to make these dictionaries be global variables. One more note, if we do item 1., then there is no need to pass sonar_model into beamgroups_to_convention (so item 2. would go away).

Great. I'm glad this is workable. 

Please see also @leewujung's alternative approach for accomplishing the same thing. I don't have a strong preference for either my or her approach. You two can decide :smiley:

leewujung commented:
I don't have a strong preference either. @b-reyes can decide :)

b-reyes commented:
> Late to the party!
> 
> On the potential blurring between `set_groups_SENSOR.py` and `set_groups_base.py`: one way to resolve this may be to just move those _definitions_ (the `beam_only_names`, `ping_time_only_names` and `beam_ping_time_names` dicts) into another script, e.g. `set_groups_dims_mod.py`. These dicts can then be used by `beamgroups_to_convention`.
> 
> This is along the same line as in our `core.py`, BUT in `core.py` the levels of organization is first with SENSOR and then with specific info, whereas this order is the reversed.

@leewujung I definitely see how this could be done and it has the upside of being very simple to implement! I think the only downside to this type of approach is that sensor specific information would be in several places. I kind of like the idea of having one place for all of the information necessary for setting the groups (analogous  to `self._beamgroups` in the sensor specific `set_groups_XX.py` files). 

Since it is up to me ... I guess I will go with @emiliom's approach as it has the upside that all of this information is in one place.

b-reyes commented:
@emiliom and @leewujung based on the above [comment](https://github.com/OSOceanAcoustics/echopype/pull/638#issuecomment-1105807694) it seems like the conclusion was that `compute_Sv` would return an xarray dataset that retains the `beam` dimension and this dimension would have length=1. 

However, after running the tests I came across a situation where the parameters `sonar_model='EK80'`, `waveform_mode='BB'`,  and `encode_mode='complex'` produce a `beam` dimension with length=4. So, it looks like we cannot just squeeze out the `beam` dimension at the [very end of `_compute_cal`](https://github.com/OSOceanAcoustics/echopype/blob/dev/echopype/calibrate/api.py#L12). 

As this seems like unexpected behavior, how should we handle this?

leewujung commented:
@b-reyes : Thanks for finding this out! The extra `beam` dimension in the outputs you saw are due to other variables being added that dimension so got carried through in the computation. I'll push up some changes.

b-reyes commented:
@emiliom and @leewujung I have addressed all of the comments you have suggested. The main portion of code that I modified was changing where the sensor-specific sets for adding the dimensions are located. This was done following @emiliom's [suggestions](https://github.com/OSOceanAcoustics/echopype/pull/638#issuecomment-1105712245), mainly: 

> 1. Move the `beam_only_names`, `ping_time_only_names` and `beam_ping_time_names` dicts to the `set_groups_SENSOR.py` modules, so they're passed as arguments to `beamgroups_to_convention`. 

Via Slack, it was decided that we will not be implementing the squeezing of `beam` in `_compute_cal` as [suggested](https://github.com/OSOceanAcoustics/echopype/pull/638#issuecomment-1105827805) in this PR because there is more discussion that is needed. This will be addressed in #644.

emiliom commented on echopype/convert/set_groups_ek60.py:
```diff
@@ -17,6 +17,36 @@
+    beam_ping_time_names are used in set_groups_base and
+    in converting from v0.5.x to v0.6.0. The values within
+    these sets are applied to all Sonar/Beam_groupX groups.
+    """
```
I'm not sure, but the presence of this string block here threw me off. It's like a docstring right after a docstring.

Also, I may be wrong here, but I don't think I've seen "stuff" placed here in a class, outside of `__init__`. I guess it works! But I would've expected it to be within `__init__`

leewujung replied:
These are "class variables" that are not instance variables under `__init__`.

leewujung replied:
Oh maybe you're just referring to the docstrings. Ha, I think there's some debate on where those go.

b-reyes replied:
Yes, I know this string looks out of place, I was using it for a multiline comment, but if you prefer, I can use #. In regards to the variables. I placed them before the `__init__` this makes them global variables of the class.

emiliom replied:
Thanks, and sorry for my ignorance there. I like @leewujung 's description -- class variables rather than instance variables.

I think I'd vote for using #'s. Otherwise my first instinct was to read that block as part of the class docstring.

b-reyes replied:
> Thanks, and sorry for my ignorance there. I like @leewujung 's description -- class variables rather than instance variables.

Yes, that is a correct and good description. 

>  I think I'd vote for using #'s. Otherwise my first instinct was to read that block as part of the class docstring.

Sounds good, unless @leewujung has any objections, I will go ahead and change this.

leewujung replied:
Go ahead and change this, and feel free to self-merge!

emiliom commented:
> I have addressed all of the comments you have suggested. The main portion of code that I modified was changing where the sensor-specific sets for adding the dimensions are located. 

Looks good, thanks! I just had one style comment.

leewujung commented:
I don't have other comments. I think this is ready to be merged!
